shm/ship.cpp

```diff
@@ -0,0 +1,173 @@
+}
+// Function returns specific cargo
+std::shared_ptr<Cargo> Ship::getCargo(size_t index) const
+{
+    if (cargos_.size() <= index)
+    {
+        return nullptr;
+    }
+    return cargos_[index];
+}
+
+std::vector<std::shared_ptr<Cargo>> Ship::getCargosVector() const
+{
+    return cargos_;
+}
+
+void Ship::setName(const std::string &name)
+{
+    name_ = name;
+}
+
+std::vector<std::shared_ptr<Cargo>>::iterator Ship::findMatchCargo(std::shared_ptr<Cargo> cargo)
+{
+    return std::find_if(cargos_.begin(),
+                        cargos_.end(),
+                        [&cargo](const auto &searchCargo)
+                        { return searchCargo->getName() == cargo->getName(); });
+}
+
+void Ship::unload(std::shared_ptr<Cargo> &cargo, size_t amount)
+{
+    auto findCargo = findMatchCargo(cargo);
+    if (findCargo == cargos_.end())
+    {
+        return;
+    }
+
+    if (findCargo != cargos_.end())
+    {
+        findCargo->get()->setAmount(findCargo->get()->getAmount() - amount);
+        if (!(*findCargo)->getAmount())
+        {
+            cargos_.erase(findCargo);
+            time_->detachObserver(dynamic_cast<Cargo *>((*findCargo).get()));
+        }
+    }
+}
+
+void Ship::load(std::shared_ptr<Cargo> &cargo, size_t amount)
+{
+    auto findCargo = findMatchCargo(cargo);
+
+    if (findCargo != cargos_.end())
+    {
+
+        if (capacity_ > cargo->getAmount())
+        {
+
+            /// I just want to know, is the line below  corrected wrote? OR should I write it, in another way?
+            findCargo->get()->setAmount(findCargo->get()->getAmount() + amount);
+        }
+        else
+        {
+            std::cerr << "Not enuogh space on ship!!!!\n";
+        }
+    }
+    else
+    {
+
+        addCargo(cargo, amount);
+    }
+}
+
+void Ship::addCargo(std::shared_ptr<Cargo> &cargo, size_t amount)
+{
+    if (Alcohol *alcohol = dynamic_cast<Alcohol *>(cargo.get()))
+    {
+        cargos_.push_back(std::make_shared<Alcohol>(alcohol->getName(),
+                                                    amount,
+                                                    alcohol->getBasePrice(),
+                                                    alcohol->getPercentage()));
+    }
+    else if (Fruit *fruit = dynamic_cast<Fruit *>(cargo.get()))
+    {
+        cargos_.push_back(std::make_shared<Fruit>(fruit->getName(),
+                                                  amount,
+                                                  fruit->getBasePrice(),
+                                                  fruit->getExpirationDate(),
+                                                  time_));
+    }
+    else if (Item *item = dynamic_cast<Item *>(cargo.get()))
+    {
+        cargos_.push_back(std::make_shared<Item>(item->getName(),
+                                                 amount,
+                                                 item->getBasePrice(),
+                                                 item->getRarity()));
+    }
```

Comment:
Dynamic cast is not good. You can avoid it by call clone() method.
cargos_push_back(cargo->clone())

Reply:
clone method will return shared_ptr and will be override by every derivered class like Fruit, Alcohol etc...